src/main/java/jenkins/plugins/git/GitToolChooser.java

```diff
@@ -0,0 +1,181 @@
+package jenkins.plugins.git;
+
+import edu.umd.cs.findbugs.annotations.NonNull;
+import hudson.EnvVars;
+import hudson.ExtensionList;
+import hudson.ExtensionPoint;
+import hudson.model.TaskListener;
+import hudson.plugins.git.GitTool;
+import hudson.plugins.git.util.GitUtils;
+import jenkins.model.Jenkins;
+import org.apache.commons.io.FileUtils;
+import org.jenkinsci.plugins.gitclient.Git;
+import org.jenkinsci.plugins.gitclient.GitClient;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.List;
+import java.util.Objects;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import java.util.stream.Collectors;
+
+/**
+ * A class which allows Git Plugin to choose a git implementation by estimating the size of a repository from a distance
+ * without requiring a local checkout.
+ */
+public class GitToolChooser {
+
+    private long sizeOfRepo = 0L;
+    private String implementation;
+    private String gitTool;
+    /**
+     * Size to switch implementation in KiB
+     */
+    public static final int SIZE_TO_SWITCH = 5000;
+
+    /**
+     * Instantiate class using {@link AbstractGitSCMSource}. It looks for a cached .git directory first, calculates the
+     * size if it is found else checks if the extension point has been implemented and asks for the size.
+     * @param source the {@link AbstractGitSCMSource}
+     * @throws IOException
+     * @throws InterruptedException
+     */
+    public GitToolChooser(@NonNull AbstractGitSCMSource source) throws IOException, InterruptedException {
+        boolean useCache;
+        boolean useAPI = false;
+
+        implementation = determineSwitchOnSize(sizeOfRepo);
+        useCache = setSizeFromCache(source);
+
+        if (useCache) {
+            implementation = determineSwitchOnSize(sizeOfRepo);
+        } else {
+            useAPI = setSizeFromAPI(source.getRemote());
+        }
+
+        if (useAPI) {
+            implementation = determineSwitchOnSize(sizeOfRepo);
+        }
+        determineGitTool(implementation);
+    }
+
+    /**
+     * Estimate size of a repository using the extension point
+     * @param remoteName: The URL of the repository
+     */
+    public GitToolChooser(String remoteName) {
+        implementation = determineSwitchOnSize(sizeOfRepo);
+
+        boolean useAPI = setSizeFromAPI(remoteName);
+        if (useAPI) {
+            implementation = determineSwitchOnSize(sizeOfRepo);
+        }
+        determineGitTool(implementation);
+    }
```

Comment:
From my comment above
```suggestion
    public GitToolChooser(String remoteName) {
        implementation = "NONE";
        useAPI();
        determineGitTool(implementation);
    }
```